### PR TITLE
node: integrate fraud service

### PR DESCRIPTION
**Pull request by vgonkivs**
The last part of #674

### CODE REVIEW

**Review by adlerjohn — APPROVED** [review]
utACK

**Review by adlerjohn — COMMENTED** [review]
misclick ignore the above 😂

File: das/daser.go [hunk 1]
```diff
@@ -70,18 +77,20 @@ func (d *DASer) Start(context.Context) error {
...
-	go d.sample(dasCtx, sub, checkpoint)
+	go d.sample(d.ctx, sub, checkpoint)
 	return nil
 }
 
 // Stop stops sampling.
 func (d *DASer) Stop(ctx context.Context) error {
+	if d.ctx.Err() != nil {
+		return nil
+	}
```
Comment (adlerjohn) [inline das/daser.go:93]: Is it okay to swallow this error silently? If yes then you should comment why, since it's not immediately obvious.

File: fraud/bad_encoding.go [hunk 2]
```diff
@@ -45,17 +45,17 @@ func CreateBadEncodingProof(
 	}
 }
 
-// Type returns type of fraud proof
+// Type returns type of fraud proof.
```
Comment (adlerjohn) [inline fraud/bad_encoding.go:48]: Do these period additions in a separate PR so as not to pollute this one with non-functional changes.

Comment (vgonkivs) [reply fraud/bad_encoding.go:48]: These changes were requested in #674, so I decided to add it here.

File: das/daser.go (hunk 1 above)
Comment (vgonkivs) [reply das/daser.go:99]: added.

File: das/daser.go [hunk 3]
```diff
@@ -70,18 +77,26 @@ func (d *DASer) Start(context.Context) error {
...
 }
 
 // Stop stops sampling.
 func (d *DASer) Stop(ctx context.Context) error {
+	// Stop func can now be invoked twice in one Lifecycle:
+	// * when BEFP is received;
+	// * when node is stopping;
+	// this statement helps to avoid panic on the second Stop.
+	// If ctx.Err is not nil then it means that Stop was called for the first time.
+	// NOTE: we are expecting here *only* ContextCancelled error.
```
Comment (Bidon15) [inline das/daser.go:96]:
```suggestion
	// NOTE: we are *only* expecting ContextCancelled error here.
```

**Review by Wondertan — COMMENTED** [review]
General notes. This PRs touches multiple pieces of stack. It also has changes unrelated to the fraud service integration, like the change of BEFP.Index type. However, all these changes are committed under the same commit. This is bad practice, and it pollutes git history, making it unreadable retroactively. Either split changes into multiple commits that touch different parts of the code, or split the PR into multiple ones. Personally, I prefer having multiple _clean_ commits within the same PR.

I am fine with keeping things like they are in this PR, considering it's long story to be merged. However, pls keep clean git history in mind for future PRs.

File: header/testing.go [hunk 4]
```diff
@@ -198,6 +201,45 @@ func RandBlockID(t *testing.T) types.BlockID {
 	return bid
 }
 
+// FraudMaker creates a custom ConstructFn that breaks the block at the given height.
+func FraudMaker(t *testing.T, faultHeight int64) ConstructFn {
+	log.Infow("Corrupting block...", "height", faultHeight)
```
Comment (Wondertan) [inline header/testing.go:206]: Warn?

File: header/testing.go [hunk 5]
```diff
@@ -198,6 +201,45 @@ func RandBlockID(t *testing.T) types.BlockID {
...
+		if b.Height == faultHeight {
+			eh = CreateFraudExtHeader(t, eh, bServ)
+			return eh, nil
+		}
+
+		namespacedShares, _, _ := b.Data.ComputeShares(uint64(0))
+		extended, err := ipld.AddShares(ctx, namespacedShares.RawShares(), bServ)
+		require.NoError(t, err)
+		dah := da.NewDataAvailabilityHeader(extended)
+		eh.DAH = &dah
```
Comment (Wondertan) [inline header/testing.go:227]: Would be nicer to reuse the code here from real HeaderMaker, but that's optional

**Review by Wondertan — COMMENTED** [review]
Mostly LGTM

File: fraud/bad_encoding.go [hunk 6]
```diff
@@ -156,10 +159,35 @@ func (p *BadEncodingProof) Validate(header *header.ExtendedHeader) error {
...
+// OnBEFP listens to Bad Encoding Fraud Proof and stops services immediately if it is received.
+func OnBEFP(ctx context.Context, s Subscriber, stop func(context.Context) error) {
+	subscription, err := s.Subscribe(BadEncoding)
+	if err != nil {
+		log.Errorw("failed to subscribe to bad encoding fraud proof", "err", err)
+		return
+	}
+	defer subscription.Cancel()
+
+	log.Info("Listening for bad encoding fraud")
```
[the remark this answers is not in the document]
Comment (vgonkivs) [reply fraud/bad_encoding.go]: removed

File: header/testing.go (hunk 5 above)
Comment (vgonkivs) [reply header/testing.go]: Yes, sounds good

File: fraud/bad_encoding.go (hunk 2 above)
Comment (vgonkivs) [reply fraud/bad_encoding.go]: all non-functional changes will be introduced in a separate PR

File: das/daser.go [hunk 7]
```diff
@@ -245,3 +261,13 @@ func (d *DASer) catchUp(ctx context.Context, job *catchUpJob) (int64, error) {
 	// report successful result
 	return job.to, nil
 }
+
+func (d *DASer) checkForByzantineError(height uint64, err error) {
+	var byzantineErr *ipld.ErrByzantine
+	if errors.As(err, &byzantineErr) {
+		err := d.bcast.Broadcast(d.ctx, fraud.CreateBadEncodingProof(height, byzantineErr))
```
[the remark this answers is not in the document]
Comment (vgonkivs) [reply das/daser.go]: done

**Comment by vgonkivs** [comment]
All non-functional changes will be added in a separate PR.

File: fraud/bad_encoding.go (hunk 2 above)
Comment (vgonkivs) [reply fraud/bad_encoding.go]: https://github.com/celestiaorg/celestia-node/pull/826

File: node/tests/fraud_test.go [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+ Pre-Requisites:
+ - CoreClient is started by swamp.
+ Steps:
+ 1. Create a Bridge Node(BN) with broken extended header at height 10.
+ 2. Start a BN.
+ 3. Create a Full Node(FN) with a connection to BN as a trusted peer.
+ 4. Start a FN.
+ 5. Subscribe to BEFP and wait when it will be received.
+ 6. Check FN is not synced to 15.
+ Note: 15 is not available because DASer will be stopped before reaching this height due to receiving BEFP.
```
[the remark this answers is not in the document]
Comment (vgonkivs) [reply node/tests/fraud_test.go:29]: added in https://github.com/celestiaorg/celestia-node/pull/826

**Comment by adlerjohn** [comment]
Can you rebase this on top of #826?

File: node/tests/fraud_test.go [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+
+	subscr, err := full.FraudServ.Subscribe(fraud.BadEncoding)
+	require.NoError(t, err)
+	_, err = subscr.Proof(ctx)
+	require.NoError(t, err)
+
+	// Since GetByHeight is a blocking operation for headers that is not received, we
+	// should set a timeout because all daser/syncer are stopped at this point
+	newCtx, cancel := context.WithTimeout(context.Background(), time.Millisecond*500)
+	t.Cleanup(cancel)
```
[the remark this answers is not in the document]
Comment (vgonkivs) [reply node/tests/fraud_test.go]: added

**Comment by vgonkivs** [comment]
#826  is based on this one. So, we should merge this first.

File: fraud/bad_encoding.go [hunk 10]
```diff
@@ -173,3 +175,27 @@ func (p *BadEncodingProof) Validate(header *header.ExtendedHeader) error {
 
 	return nil
 }
+
+// OnBEFP listens to Bad Encoding Fraud Proof and stops services immediately if it is received.
+func OnBEFP(ctx context.Context, s Subscriber, stop func(context.Context) error) {
+	subscription, err := s.Subscribe(BadEncoding)
+	if err != nil {
+		log.Errorw("failed to subscribe to bad encoding fraud proof", "err", err)
```
Comment (renaynay) [inline fraud/bad_encoding.go:183]: If this actually occurs, the node shouldn't start. We shouldn't have/allow nodes running on the network that aren't subscribed to BEFP.

Comment (renaynay) [reply fraud/bad_encoding.go:183]: We should separate the subscription step from the listening for Proof step somehow

File: fraud/testing.go [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+}
+
+func (d *dummySubscription) Proof(ctx context.Context) (Proof, error) {
+	<-ctx.Done()
+	return nil, ctx.Err()
+}
+
+func (d *dummySubscription) Cancel() {}
+
+type mockStore struct {
```
Comment (renaynay) [inline fraud/testing.go:50]: we also have mockStore in `exchange_test`, maybe we can extract into its own testing file? This can be done in a later PR, otherwise we have duplicate code here

Comment (renaynay) [reply fraud/testing.go:50]: Can you make an issue for this? and link to this PR

File: header/sync/sync.go [hunk 12]
```diff
@@ -59,10 +58,6 @@ func NewSyncer(exchange header.Exchange, store header.Store, sub header.Subscrib
 
 // Start starts the syncing routine.
 func (s *Syncer) Start(context.Context) error {
-	if s.cancel != nil {
```
Comment (renaynay) [inline header/sync/sync.go:62]: Why remove?

File: header/sync/sync.go [hunk 13]
```diff
@@ -76,10 +71,9 @@ func (s *Syncer) Start(context.Context) error {
 }
 
 // Stop stops Syncer.
-func (s *Syncer) Stop(context.Context) error {
+func (s *Syncer) Stop(ctx context.Context) error {
 	s.cancel()
-	s.cancel = nil
```
Comment (renaynay) [inline header/sync/sync.go:81]: why?

File: node/services/service.go [hunk 14]
```diff
@@ -131,20 +136,32 @@ func ShareService(lc fx.Lifecycle, bServ blockservice.BlockService, avail share.
...
+	das := das.NewDASer(avail, sub, hstore, ds, fservice)
 	lc.Append(fx.Hook{
 		OnStart: das.Start,
 		OnStop:  das.Stop,
 	})
+	go fraud.OnBEFP(fxutil.WithLifecycle(ctx, lc), fservice, das.Stop)
 	return das
 }
 
+// FraudService constructs fraud proof service.
```
Comment (renaynay) [inline node/services/service.go:156]:
```suggestion
// FraudService constructs fraud proof service for bad encoding fraud proofs (BEFPs).
```

File: service/rpc/middleware.go [hunk 15]
```diff
@@ -14,3 +20,17 @@ func setContentType(next http.Handler) http.Handler {
 		next.ServeHTTP(w, r)
 	})
 }
+
+// checkAllowance ensures that context was canceled and prohibit POST requests.
+func checkAllowance(state *state.Service) mux.MiddlewareFunc {
```
Comment (renaynay) [inline service/rpc/middleware.go:25]: Nit but we should call this `checkPostDisabled`

Comment (renaynay) [reply service/rpc/middleware.go:25]: `allowance` has a different meaning :)

File: service/rpc/middleware.go [hunk 16]
```diff
@@ -14,3 +20,17 @@ func setContentType(next http.Handler) http.Handler {
...
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			// check if state service was halted and deny the transaction
```
Comment (renaynay) [inline service/rpc/middleware.go:28]:
```suggestion
			// check if state service was halted and deny the POST request
```

**Review by renaynay — COMMENTED** [review]
Minor comments.

File: header/sync/sync.go (hunk 13 above)
Comment (vgonkivs) [reply header/sync/sync.go:81]: Syncer can be stopped twice when:
1) BEFP is received;
2) node is stopping;

To avoid any potential races we agreed to remove ` if s.cancel != nil {`. Now cancel can be invoked twice but it's not a problem.

File: das/daser.go [hunk 17]
```diff
@@ -70,18 +77,26 @@ func (d *DASer) Start(context.Context) error {
...
 
 // Stop stops sampling.
 func (d *DASer) Stop(ctx context.Context) error {
+	// Stop func can now be invoked twice in one Lifecycle, when:
+	// * BEFP is received;
+	// * node is stopping;
+	// this statement helps avoiding panic on the second Stop.
+	// If ctx.Err is not nil then it means that Stop was called for the first time.
+	// NOTE: we are expecting *only* ContextCancelled error here.
+	if d.ctx.Err() != nil {
```
[the remark this answers is not in the document]
Comment (vgonkivs) [reply das/daser.go:97]: We do not expect any other errors except context.Cancelled but ok.

File: header/sync/sync.go (hunk 12 above)
Comment (vgonkivs) [reply header/sync/sync.go:62]: To avoid races. Please, check next comment.

File: service/rpc/middleware.go (hunk 16 above)
Comment (vgonkivs) [reply service/rpc/middleware.go:28]: then may be `POST requests`?

File: das/daser.go [hunk 18]
```diff
@@ -245,3 +262,13 @@ func (d *DASer) catchUp(ctx context.Context, job *catchUpJob) (int64, error) {
 	// report successful result
 	return job.to, nil
 }
+
+func (d *DASer) checkForByzantineError(ctx context.Context, height uint64, err error) {
+	var byzantineErr *ipld.ErrByzantine
+	if errors.As(err, &byzantineErr) {
+		err := d.bcast.Broadcast(ctx, fraud.CreateBadEncodingProof([]byte("hash"), height, byzantineErr))
```
[the remark this answers is not in the document]
Comment (vgonkivs) [reply das/daser.go:269]: instead of error?

File: fraud/bad_encoding.go (hunk 10 above)
Comment (vgonkivs) [reply fraud/bad_encoding.go:183]: We should keep in mind if Subscribe failed the we have to call `stop()`. So Subscribe+Listen should be somewhere in one place.
We can implement smth similar to `fraudSession` - a custom struct that will keep subscription + stop function and it will call stop once Fraud proof will be received.

cc @Wondertan , @liamsi

File: fraud/testing.go (hunk 11 above)
Comment (vgonkivs) [reply fraud/testing.go:50]: https://github.com/celestiaorg/celestia-node/issues/853

File: header/sync/sync.go (hunk 13 above)
Comment (renaynay) [reply header/sync/sync.go:81]: ack

File: header/sync/sync.go (hunk 12 above)
Comment (renaynay) [reply header/sync/sync.go:62]: ack

File: fraud/testing.go (hunk 11 above)
Comment (renaynay) [reply fraud/testing.go:50]: ack

File: fraud/bad_encoding.go (hunk 10 above)
Comment (renaynay) [reply fraud/bad_encoding.go:190]: If we have subscription construction happen separately, and for some reason, that fails, fx will fail out and not allow the node to start.

Comment (Wondertan) [reply fraud/bad_encoding.go:190]: Ok, just saw that @vgonkivs already implemented something similar

File: fraud/bad_encoding.go [hunk 19]
```diff
@@ -173,3 +175,31 @@ func (p *BadEncodingProof) Validate(header *header.ExtendedHeader) error {
...
+	}
+	defer subscription.Cancel()
+
+	// At this point we receive already verified fraud proof,
+	// so there are no needs to call Validate.
+	_, err = subscription.Proof(ctx)
+	if err != nil {
+		if err == context.Canceled {
+			return
+		}
```
Comment (Wondertan) [inline fraud/bad_encoding.go:201]: nit: This can be simplified into one if.

File: fraud/bad_encoding.go [hunk 20]
```diff
@@ -173,3 +175,31 @@ func (p *BadEncodingProof) Validate(header *header.ExtendedHeader) error {
...
+// OnBEFP listens to Bad Encoding Fraud Proof and stops services immediately if it is received.
+func OnBEFP(ctx context.Context, s Subscriber, stop func(context.Context) error) {
+	subscription, err := s.Subscribe(BadEncoding)
+	defer func() {
+		if err != context.Canceled {
+			if stopErr := stop(ctx); err != nil {
+				log.Warn(stopErr)
+			}
+		}
+	}()
```
Comment (Wondertan) [inline fraud/bad_encoding.go:188]:
```suggestion
func OnBEFP(ctx context.Context, s Subscriber, stop func(context.Context) (err error) {
	defer func() {
		if err == context.Canceled {
			return
		}
		
		stopErr := stop(ctx)
		if stopErr != nil {
			log.Warn(stopErr)
		}
	}()
	
	subscription, err := s.Subscribe(BadEncoding)
```

Comment (Wondertan) [reply fraud/bad_encoding.go:188]: * This is a good case to define a global named return err and make defer for it.
* Don't do nested if
* there was a bug where stopErr was not checked in if before log.Warn

**Review by Wondertan — APPROVED** [review]
LGTM

File: fraud/bad_encoding.go (hunk 10 above)
Comment (renaynay) [reply fraud/bad_encoding.go:190]: ack

**Review by renaynay — APPROVED** [review]

**Review by Bidon15 — APPROVED** [review]
Awesome 🎬

**Review by Wondertan — APPROVED** [review]

**Comment by Wondertan** [comment]
One thing though. Let's not merge this until https://github.com/celestiaorg/celestia-node/pull/849 settles.

This way we can avoid ugly `checkForByzantineError` and call it once inside `sampleHeader` introduced in #849

**Review by adlerjohn — APPROVED** [review]
utACK

**Comment by Wondertan** [comment]
@vgonkivs, the blocking PR is merged, pls rebase and we are ready to go

File: fraud/bad_encoding.go [hunk 21]
```diff
@@ -99,6 +99,8 @@ func (p *BadEncodingProof) UnmarshalBinary(data []byte) error {
 		headerHash:  in.HeaderHash,
 		BlockHeight: in.Height,
 		Shares:      ipld.ProtoToShare(in.Shares),
+		Index:       uint8(in.Index),
+		isRow:       in.IsRow,
```
Comment (Wondertan) [inline fraud/bad_encoding.go:103]: Is not this supposed to be part of findings Pr?

File: das/daser.go [hunk 22]
```diff
@@ -240,3 +257,15 @@ func (d *DASer) sampleHeader(ctx context.Context, h *header.ExtendedHeader) erro
 
 	return nil
 }
+
+func (d *DASer) checkForByzantineError(ctx context.Context, hash []byte, height uint64, err error) {
```
[the remark this answers is not in the document]
Comment (vgonkivs) [reply das/daser.go:260]: yes, currently reworking it.

File: fraud/bad_encoding.go (hunk 21 above)
Comment (vgonkivs) [reply fraud/bad_encoding.go:103]: wdym?

Comment (vgonkivs) [reply fraud/bad_encoding.go:103]: Ah, got you. Befp won't work correctly without this 😅

Comment (Wondertan) [reply fraud/bad_encoding.go:103]: 👍

**Review by Wondertan — APPROVED** [review]

File: das/daser.go [hunk 23]
```diff
@@ -264,6 +263,8 @@ func (d *DASer) checkForByzantineError(ctx context.Context, hash []byte, height
 		err := d.bcast.Broadcast(ctx, fraud.CreateBadEncodingProof(hash, height, byzantineErr))
 		if err != nil {
 			log.Errorw("fraud proof propagating failed", "err", err)
+			return
 		}
+		log.Warn("fraud detected and broadcasted successfully. Shutting down services…")
```
Comment (renaynay) [inline das/daser.go:268]: I would log.Warn("Shutting down services...") regardless of whether there was an error broadcasting or not, no?

**Review by renaynay — APPROVED** [review]

File: das/daser.go [hunk 24]
```diff
@@ -242,7 +242,15 @@ func (d *DASer) sampleHeader(ctx context.Context, h *header.ExtendedHeader) erro
...
 		}
-		d.checkForByzantineError(d.ctx, h.Hash(), uint64(h.Height), err)
+		var byzantineErr *ipld.ErrByzantine
+		if errors.As(err, &byzantineErr) {
+			sendErr := d.bcast.Broadcast(ctx, fraud.CreateBadEncodingProof(h.Hash(), uint64(h.Height), byzantineErr))
+			if sendErr != nil {
+				log.Errorw("fraud proof propagating failed", "err", err)
+			} else {
+				log.Warn("Shutting down services…")
+			}
```
Comment (renaynay) [inline das/daser.go:252]:
```suggestion
                         log.Warn("Bad encoding fraud detected, propagating proof...")
			sendErr := d.bcast.Broadcast(ctx, fraud.CreateBadEncodingProof(h.Hash(), uint64(h.Height), byzantineErr))
			if sendErr != nil {
				log.Errorw("fraud proof propagating failed", "err", err)
			} 
			log.Warn("Shutting down services…")

```

**Review by Bidon15 — APPROVED** [review]

File: das/daser.go (hunk 24 above)
Comment (vgonkivs) [reply das/daser.go:252]: I'd suggest to leave only `log.Warn("Propagating proof...")`

Comment (Wondertan) [reply das/daser.go:252]: If we are talking about logs here. I don't think "Shutting down services…" log is appropriate to be in DASer, while it should be either in the OnBefp or FraudService, or just every Service should log that it is shutting down on Stop

Comment (vgonkivs) [reply das/daser.go]: agree. please take a look https://github.com/celestiaorg/celestia-node/pull/820/commits/f38abcac485c7ada59d9358de7f9e4630b0bd411
